suro-kafka-producer/src/main/java/com/netflix/suro/sink/kafka/KafkaSink.java

```diff
@@ -215,49 +216,42 @@ private void dropMessage(MessageContainer message) {
     public void open() {
         producer = new KafkaProducer(props);
         subscription = stream
-            .filter(new Func1<MessageContainer, Boolean>() {
-                @Override
-                public Boolean call(MessageContainer message) {
-                    if (!metadataWaitingQueuePolicy.acquire()) {
-                        dropMessage(message);
-                        return false;
-                    } else {
-                        return true;
-                    }
-                }
-            })
-            .subscribe(
-                new Action1<MessageContainer>() {
+                .filter(new Func1<MessageContainer, Boolean>() {
                     @Override
-                    public void call(final MessageContainer message) {
-                        Runnable r = new Runnable() {
-
+                    public Boolean call(MessageContainer message) {
+                        if (!metadataWaitingQueuePolicy.acquire()) {
+                            dropMessage(message);
+                            return false;
+                        } else {
+                            return true;
+                        }
+                    }
+                })
+                .observeOn(Schedulers.from(executor))
+                .subscribe(
+                        new Action1<MessageContainer>() {
                             @Override
-                            public void run() {
+                            public void call(final MessageContainer message) {
                                 try {
+                                    metadataWaitingQueuePolicy.release();
```

Comment:
I saw your comment about this before but I forgot to respond.

producer.partitionsFor() is blocking code so, release() before fetching metadata can cause trouble, I guess.